EngineLayer/Proteomics/Protease.cs

```diff
@@ -87,9 +93,337 @@ internal List<int> GetDigestionSiteIndices(string sequence)
+            {
+                intervals.AddRange(SemiProteolyticDigestion(protein, initiatorMethionineBehavior, maximumMissedCleavages, minPeptidesLength, maxPeptidesLength));
+            }
+
+            // Full proteolytic cleavage
+            else if (CleavageSpecificity == CleavageSpecificity.Full)
+            {
+                intervals.AddRange(FullDigestion(protein, initiatorMethionineBehavior, maximumMissedCleavages, minPeptidesLength, maxPeptidesLength));
+            }
+
+            else
+            {
+                throw new NotImplementedException();
+            }
+
+            return intervals;
+        }
+
+        /// <summary>
+        /// Gets the protein intervals based on nonspecific digestion rules
+        /// </summary>
+        /// <param name="protein"></param>
+        /// <param name="initiatorMethionineBehavior"></param>
+        /// <param name="maximumMissedCleavages"></param>
+        /// <param name="minPeptidesLength"></param>
+        /// <param name="maxPeptidesLength"></param>
+        /// <returns></returns>
+        private IEnumerable<ProteinInterval> SemiProteolyticDigestion(Protein protein, InitiatorMethionineBehavior initiatorMethionineBehavior, int maximumMissedCleavages, int? minPeptidesLength, int? maxPeptidesLength)
+        {
+            List<ProteinInterval> intervals = new List<ProteinInterval>();
+            List<int> oneBasedIndicesToCleaveAfter = GetDigestionSiteIndices(protein.BaseSequence);
+            for (int i = 0; i < oneBasedIndicesToCleaveAfter.Count - maximumMissedCleavages - 1; i++)
+            {
+                bool retain = Retain(i, initiatorMethionineBehavior, protein[0]);
+                bool cleave = Cleave(i, initiatorMethionineBehavior, protein[0]);
+                if (!retain && !cleave)
+                {
+                    continue;
+                }
+
+                int nTerminusProtein = retain ? oneBasedIndicesToCleaveAfter[i] : 1;
+                int cTerminusProtein = oneBasedIndicesToCleaveAfter[i + maximumMissedCleavages + 1];
+                if (OkayLength(cTerminusProtein - nTerminusProtein, minPeptidesLength, maxPeptidesLength))
+                {
+                    intervals.Add(new ProteinInterval(nTerminusProtein + 1, cTerminusProtein, protein, cTerminusProtein - nTerminusProtein, "semi" + (cleave ? ":M cleaved" : ""))); // Maximum sequence length
+                }
+
+                // Fixed termini at each internal index
+                IEnumerable<int> internalIndices = Enumerable.Range(nTerminusProtein - 1, cTerminusProtein - nTerminusProtein - 1);
+
+                IEnumerable<ProteinInterval> fixedCTermIntervals = internalIndices.Where(j => OkayLength(cTerminusProtein - j, minPeptidesLength, maxPeptidesLength))
+                    .Select(j => new ProteinInterval(j + 1, cTerminusProtein, protein, cTerminusProtein - j, "semiC" + (cleave ? ":M cleaved" : "")));
```

Comment:
@zrolfs I think you may have forgotten to append ":M cleaved" to these "semi" peptide strings